connection.go

```diff
@@ -1382,49 +1385,73 @@ func (c *Connection) removeAckedFrames(pn uint64, qp *[]frame) {
 	*qp = q
 }
 
+func (c* Connection) processAckRange(start uint64, end uint64, protected bool){
+	pn := start
+	// Unusual loop structure to avoid weirdness at 2^64-1
+	for {
+		// TODO(ekr@rtfm.com): properly filter for ACKed packets which are in the
+		// wrong key phase.
+		c.log(logTypeConnection, "%s: processing ACK for PN=%x", c.label(), pn)
+
+		// 1. Go through the outgoing queues and remove all the acked chunks.
+		c.removeAckedFrames(pn, &c.outputClearQ)
+		if protected {
+			c.removeAckedFrames(pn, &c.outputProtectedQ)
+		}
+
+		// 2. Mark all the packets that were ACKed in this packet as double-acked.
+		acks, ok := c.sentAcks[pn]
+		if ok {
+			for _, a := range acks {
+				c.log(logTypeAck, "Ack2 for ack range last=%v len=%v", a.lastPacket, a.count)
+				for i := uint64(0); i < a.count; i++ {
+					c.recvd.packetSetAcked2(a.lastPacket - i)
+				}
+			}
+		}
+
+		if pn == end {
+			break
+		}
+		pn++
+	}
+}
+
 func (c *Connection) processAckFrame(f *ackFrame, protected bool) error {
 	end := f.LargestAcknowledged
 	start := end - f.AckBlockLength
 
-	// Go through all the ACK blocks and process everything.
-	for {
-		c.log(logTypeAck, "%s: processing ACK range %x-%x", c.label(), start, end)
-		// Unusual loop structure to avoid weirdness at 2^64-1
-		pn := start
-		for {
-			// TODO(ekr@rtfm.com): properly filter for ACKed packets which are in the
-			// wrong key phase.
-			c.log(logTypeConnection, "%s: processing ACK for PN=%x", c.label(), pn)
-
-			// 1. Go through the outgoing queues and remove all the acked chunks.
-			c.removeAckedFrames(pn, &c.outputClearQ)
-			if protected {
-				c.removeAckedFrames(pn, &c.outputProtectedQ)
-			}
+	/* Process the First ACK Block */
+	c.log(logTypeAck, "%s: processing ACK range %x-%x", c.label(), start, end)
+	c.processAckRange(start, end, protected)
+
+	/* Process aditional ACK Blocks */
+	last := start
+	rawAckBlocks := f.AckBlockSection
+	assert(len(rawAckBlocks) == int(f.NumBlocks * 5)) //TODO manage non 32-bit ack blocks
```

Comment:
Please mark TODO with the name of the person responsible. Your email if you are going to, else <ekr@rtfm.com>